File: citadel-drone/src/lib/hive.ts
import makeLogger from './logger';
import { IJob } from '@citadelnest/lib';
import dispatcher from '../dispatcher';

const logger = makeLogger(module);

let HOST: string, TOKEN: string;

const queryHive = async <T>(
  endpoint: string,
  method: 'POST' | 'PUT' | 'GET' | 'PATCH' | 'DELETE',
  data?: unknown
): Promise<T> => {
  const response = await fetch(new URL(endpoint, HOST), {
    headers: { 'content-type': 'application/json;charset=UTF-8', 'X-Api-Key': TOKEN },
    method,
    body: JSON.stringify(data),
  });

  if (response.ok) return { ...(await response.json()), status: response.status };

  if (response.status < 500) throw new Error((await response.json()).message);

  throw new Error('Unhandled error from hive');
};

const connectToHive = async (host: string, token: string) => {
  try {
    HOST = host;
    TOKEN = token;
    await queryHive('/drones/register', 'POST', {});
  } catch (e) {
    throw new Error(`Cannot connect to the hive (${e.cause ?? e.message})`);
  }
};

const pollNextJob = async () => {
  const { status, data } = await queryHive<{ status: number; data: { job: IJob } }>('/jobs', 'GET');

  if (status === 201) return;
  if (data?.job === undefined || data.job === null) return;

  const { job } = data;

  logger.info(`New job polled (${job.jobType}) (${job.status})`);

  dispatcher(job);
};

const tryPollNextJob = async () => {
  try {
    await pollNextJob();
  } catch (e) {
    logger.error(`Error while polling next job: ${e}`);
  }
};

export { connectToHive, pollNextJob, tryPollNextJob, queryHive };

File: citadel-drone/src/hive.rs
use std::sync::OnceLock;

use log::info;
use serde::{de::DeserializeOwned, Deserialize, Serialize};

use crate::{job, providers};

#[derive(Debug)]
pub struct Hive {
    url: String,
    token: String,
}
static INSTANCE: OnceLock<Hive> = OnceLock::new();

#[derive(Debug)]
pub enum HiveError {
    ReqwestError(reqwest::Error),
    SerdeError(serde_json::Error),
}

impl From<reqwest::Error> for HiveError {
    fn from(value: reqwest::Error) -> Self {
        Self::ReqwestError(value)
    }
}

impl From<serde_json::Error> for HiveError {
    fn from(value: serde_json::Error) -> Self {
        Self::SerdeError(value)
    }
}

#[derive(Debug, Deserialize)]
pub struct HiveResponse<T> {
    pub error: bool,
    pub data: T,
}

#[derive(Debug, Deserialize)]
pub struct DroneRegisterResponseDrone {
    name: String,
}

#[derive(Debug, Deserialize)]
pub struct DroneRegisterResponse {
    drone: DroneRegisterResponseDrone,
}

#[derive(Debug, Deserialize)]
pub struct JobResponse {
    job: Option<job::Job>,
}

impl Hive {
    pub async fn initialize(
        url: String,
        token: String,
    ) -> Result<HiveResponse<DroneRegisterResponse>, HiveError> {
        INSTANCE
            .set(Self { url, token })
            .expect("Cannot initialize Hive");

        Self::query::<DroneRegisterResponse, ()>("/drones/register", reqwest::Method::POST, None)
            .await
    }

    async fn build_client(hive: &Hive) -> Result<reqwest::Client, HiveError> {
        let mut headers = reqwest::header::HeaderMap::new();
        headers.insert(
            reqwest::header::CONTENT_TYPE,
            reqwest::header::HeaderValue::from_static("application/json;charset=UTF-8"),
        );

        headers.insert(
            "X-Api-Key",
            reqwest::header::HeaderValue::from_str(hive.token.as_str()).unwrap(),
        );

        let client_builder = reqwest::Client::builder();
        let client = client_builder.default_headers(headers).build()?;

        Ok(client)
    }

    pub async fn query<T, B>(
        endpoint: &str,
        method: reqwest::Method,
        body: Option<&B>,
    ) -> Result<HiveResponse<T>, HiveError>
    where
        T: DeserializeOwned,
        B: Serialize,
    {
        let hive = INSTANCE.get().expect("Hive must be initialized");
        let client = Self::build_client(&hive).await?;

        let mut request = client.request(method, format!("{}{}", hive.url, endpoint));

        if body.is_some() {
            request = request.json::<B>(&body.unwrap());
        }

        /* let request2 = request.try_clone().unwrap().send().await?.text().await;
        println!("{request2:?}"); */

        let result = request.send().await?;

        Ok(result.json::<HiveResponse<T>>().await?)
    }

    pub async fn fetch_jobs(provider: &impl providers::ProviderImpl) -> () {
        let job = Hive::query::<JobResponse, ()>("/drone/jobs", reqwest::Method::GET, None).await;

        if job.is_err() {
            info!("{job:?}");

            return;
        }

        let unwrapped_job_option = job.unwrap();

        if unwrapped_job_option.data.job.is_none() {
            return;
        }

        let final_job = unwrapped_job_option.data.job.unwrap();

        final_job.execute(provider).await;
    }
}
